drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,319 @@
+			priority++
+			_, taken = takenPriorities[priority]
+		}
+		rule.Priority = to.Int32Ptr(priority)
+		priority++
+		newRules = append(newRules, rule)
+	}
+	return &newRules, nil
+}
+
+// checkIfDockerConflict does a rudimentary check to see if the rule could potentially conflict with the ability
+// for the docker port to be accessible. If the rule does not pass all rudimentary checks, it returns an error.
+func checkIfDockerConflict(rule network.SecurityRule, dockerPort int) error {
+	f := logutil.Fields{
+		"name":        to.String(rule.Name),
+		"description": to.String(rule.Description),
+		"priority":    to.Int32(rule.Priority),
+		"access":      rule.Access,
+		"direction":   rule.Direction,
+		"protocol":    rule.Protocol,
+		"portRanges":  getDestinationPorts(rule),
+	}
+	priority := to.Int32(rule.Priority)
+	if priority > dockerPriority {
+		return nil
+	}
+	purpose := strings.ToLower(fmt.Sprintf("%s %s %s", rule.Access, rule.Direction, rule.Protocol))
+	if purpose == "deny inbound tcp" || purpose == "deny inbound *" {
+		// check if docker port is covered by rule
+		for _, pr := range getDestinationPorts(rule) {
+			portWithinRange, err := checkIfPortInPortRange(dockerPort, pr)
+			if err != nil {
+				return fmt.Errorf("%s: %s", err, f)
+			}
+			if portWithinRange {
+				return fmt.Errorf("rule %s set to %s requests to the Docker port %d: %s",
+					to.String(rule.Name), purpose, dockerPort, f)
+			}
+		}
+	}
+	return nil
+}
+
+func checkIfPortInPortRange(port int, portRange string) (bool, error) {
+	if portRange == "*" {
+		return true, nil
+	}
+	s := strings.Split(portRange, "-")
+	if len(s) == 2 {
+		// Port range represents is a range i.e. "2000-3000"
+		portMin, portMinErr := strconv.Atoi(s[0])
+		portMax, portMaxErr := strconv.Atoi(s[1])
+		if portMinErr == nil || portMaxErr == nil || portMin < portMax {
+			return (port >= portMin) && (port <= portMax), nil
+		}
+	}
+	if len(s) == 1 {
+		// Port range represents a single port i.e. "2376"
+		portVal, err := strconv.Atoi(s[0])
+		if err == nil {
+			return port == portVal, nil
+		}
+	}
+	return false, fmt.Errorf("invalid port range %s", portRange)
+}
+
+func getDestinationPorts(rule network.SecurityRule) []string {
+	ports := []string{}
+	if rule.DestinationPortRange != nil {
+		ports = append(ports, *rule.DestinationPortRange)
+	}
+	if rule.DestinationPortRanges != nil {
+		for _, port := range *rule.DestinationPortRanges {
+			ports = append(ports, port)
+		}
+	}
+	return ports
+}
+
+func removeSecurityRule(rules *[]network.SecurityRule, rule network.SecurityRule) *[]network.SecurityRule {
+	newRules := *rules
+	for i, elem := range newRules {
+		if elem != rule {
+			continue
+		}
+		if len(newRules) == 1 {
+			return &[]network.SecurityRule{}
+		}
+		lastElemIndex := len(newRules) - 1
+		newRules[i] = newRules[lastElemIndex]
+		newRules = newRules[:lastElemIndex]
+	}
+	return &newRules
+}
+
+// OpenPort is a port to be opened within an NSG
+type OpenPort struct {
```

Comment:
does this need to be exported? feel it's used just in this file

Reply:
It's just used in this one file; makes sense to unexport it.